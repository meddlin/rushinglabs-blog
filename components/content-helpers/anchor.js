const Anchor = ({ children, href }) => {
    return (
        <a href={href} 
            className="
                underline 
                underline-offset-1
                decoration-solid 
                decoration-slate-400
                decoration-1
                hover:decoration-indigo-500
                hover:decoration-2">
            {children}
        </a>
    );
}

export default Anchor;