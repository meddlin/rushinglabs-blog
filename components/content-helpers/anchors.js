import styles from './anchors.module.css';

function getAnchor(text) {
    return text
        .toLowerCase()
        .replace(/[^a-z0-9 ]/g, '')
        .replace(/[ ]/g, '-');
}

const H2 = ({ children }) => {
    const anchor = getAnchor(children);
    const link = `#${anchor}`;
    return (
        <h2 id={anchor} style={{ position: `relative` }}>
            {children}
            <a href={link} className={styles.anchor}>
                ยง
            </a>
        </h2>
    )
}

const H3 = ({ children }) => {
    const anchor = getAnchor(children);
    const link = `#${anchor}`;
    return (
        <h3 id={anchor} style={{ position: `relative` }}>
            {children}
            <a href={link} className={styles.anchor}>
                ยง
            </a>
        </h3>
    )
}

const H4 = ({ children }) => {
    const anchor = getAnchor(children);
    const link = `#${anchor}`;
    return (
        <h4 id={anchor} style={{ position: `relative` }}>
            {children}
            <a href={link} className={styles.anchor}>
                ยง
            </a>
        </h4>
    )
}

export { H2, H3, H4, getAnchor };