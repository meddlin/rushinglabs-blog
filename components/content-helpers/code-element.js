const CodeElement = ({ children }) => {
    return (
        <code 
            className="
                text-xs
                bg-slate-200
                rounded
                px-px
            ">
            {children}
        </code>
    )
}

export default CodeElement;