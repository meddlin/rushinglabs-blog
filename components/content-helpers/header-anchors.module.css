.anchor {
    color: #666;
    opacity: 0;
    position: absolute;
    transform: translate(0.25em, -2px);
    width: 1em;
}

h2:hover .anchor {
    opacity: 1;
}
h3:hover .anchor {
    opacity: 1;
}
h4:hover .anchor {
    opacity: 1;
}