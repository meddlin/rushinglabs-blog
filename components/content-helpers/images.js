import Image from 'next/image';

/**
 * ImageContainer - wraps the next/image component to provide standardized use/placement of alt, credits
 * - Defaults to a center image placement
 * - Places caption, image credit, and image credit link directly below the image
 * - ImageContainer is not exported because it is only supposed to be an internal utility to these image 
 *      management components
 */
function ImageContainer({ children, src, height, width, alt, caption, imageCredit, imageCreditLink, placeholder, blurDataURL }) {
    return (
        <div style={{ display: `flex`, justifyContent: `center`, alignItems: `center`, flexDirection: `column` }}>
            <div style={{ filter: `drop-shadow(0.15rem 0.25rem 0.45rem darkgrey)` }}>
                <div style={{ filter: `drop-shadow(0.15rem 0.25rem 0.45rem darkgrey)` }}>
                    <Image src={src}
                        height={height}
                        width={width}
                        alt={alt}
                        placeholder={placeholder}
                        blurDataURL={blurDataURL} />
                </div>
                {caption ? (<p>
                    <i>
                        {caption}
                    </i>
                </p>) : ''}
                {imageCredit || imageCreditLink ? (
                    <p>
                        <i>Credit: <a href={imageCreditLink}>{imageCredit}</a></i>
                    </p>
                ) : ''}
            </div>
        </div>
    );
}

/**
 * CenteredImage
 * - An image component for center-aligned images, and no slots for children content. 
 * - All caption and image credit information is rendered directly beneath the image, center-aligned too.
 * 
 * @param {*} param0 
 * @returns 
 */
function CenteredImage({ src, height, width, alt, caption, imageCredit, imageCreditLink, placeholder, blurDataURL }) {
    return (
        <ImageContainer
            src={src}
            height={height}
            width={width}
            placeholder={placeholder}
            blurDataURL={blurDataURL}

            alt={alt}
            caption={caption}
            imageCredit={imageCredit}
            imageCreditLink={imageCreditLink}
        />
    );
}

/**
 * LeftImage
 * - An image component with left-aligned image, and content to the right.
 * @param {*} param0 
 * @returns 
 */
function LeftImage({ children, src, height, width, alt, caption, imageCredit, imageCreditLink, placeholder, blurDataURL }) {
    return (
        <div style={{ display: `flex`, justifyContent: `flex-start`, alignItems: `center`, flexDirection: `row` }}>
            <ImageContainer
                src={src}
                height={height}
                width={width}
                placeholder={placeholder}
                blurDataURL={blurDataURL}

                alt={alt}
                caption={caption}
                imageCredit={imageCredit}
                imageCreditLink={imageCreditLink}
            />

            <div style={{ minWidth: `30%`, marginLeft: `1em` }}>
                {children}
            </div>
        </div>
    );
}

/**
 * RightImage
 * - An image component with right-aligned image, and content to the left.
 * @param {*} param0 
 * @returns 
 */
function RightImage({ children, src, height, width, alt, caption, imageCredit, imageCreditLink, placeholder, blurDataURL }) {
    return (
        <div style={{ display: `flex`, justifyContent: `flex-end`, alignItems: `center`, flexDirection: `row` }}>
            <div style={{ minWidth: `30%`, marginRight: `1em` }}>
                {children}
            </div>
            <ImageContainer
                src={src}
                height={height}
                width={width}
                placeholder={placeholder}
                blurDataURL={blurDataURL}

                alt={alt}
                caption={caption}
                imageCredit={imageCredit}
                imageCreditLink={imageCreditLink}
            />
        </div>
    );
}

export { CenteredImage, LeftImage, RightImage };