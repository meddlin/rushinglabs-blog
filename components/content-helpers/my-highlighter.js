import SyntaxHighlighter from 'react-syntax-highlighter';
import { docco, a11yDark } from 'react-syntax-highlighter/dist/cjs/styles/hljs'

const MyHighlighter = ({ 
    language, 
    children, 
    showLineNumbers, 
    showInlineLineNumbers 
}) => {
    return (
        <SyntaxHighlighter 
            className="text-xs" // Tailwind classes here

            language={language} 
            style={a11yDark} 
            showLineNumbers={showLineNumbers ? true : false}
            showInlineLineNumbers={showInlineLineNumbers ? true : false}
        >
                {children}
        </SyntaxHighlighter>
    );
}

export default MyHighlighter;