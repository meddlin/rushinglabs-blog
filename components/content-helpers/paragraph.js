const Paragraph = ({ children }) => {
    return (
        <p className="my-2">
            {children}
        </p>
    );
}

export default Paragraph;