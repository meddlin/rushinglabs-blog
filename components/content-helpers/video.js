/**
 * Video
 * - Takes in a .webm video, and creates an HTML5 WebM embedded video for it.
 * - Can optionally pass in a width for the video
 */
function Video({ children, src, width }) {
    // TODO: Test the 'src' prop for having .webm as a file suffix
    width = width !== undefined || "" ? width : "100%";

    return (
        <video controls width={width}>
            <source src={src}
                type="video/webm" />
        </video>
    );
}

export { Video };