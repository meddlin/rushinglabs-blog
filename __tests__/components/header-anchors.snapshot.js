import { render } from '@testing-library/react';
import { H2, H3, H4 } from '../../components/content-helpers/header-anchors';

it('renders H2 unchanged', () => {
    const { container } = render(<H2>this is a test</H2>);

    expect(container).toMatchSnapshot()
});

it('renders H3 unchanged', () => {
    const { container } = render(<H3>this is a test</H3>);

    expect(container).toMatchSnapshot()
});

it('renders H4 unchanged', () => {
    const { container } = render(<H4>this is a test</H4>);

    expect(container).toMatchSnapshot()
});