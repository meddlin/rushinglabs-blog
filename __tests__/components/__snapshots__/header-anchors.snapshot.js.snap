// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders H2 unchanged 1`] = `
<div>
  <h2
    class="relative text-xl mt-6"
    id="this-is-a-test"
  >
    this is a test
    <a
      class="anchor"
      href="#this-is-a-test"
    >
      ยง
    </a>
  </h2>
</div>
`;

exports[`renders H3 unchanged 1`] = `
<div>
  <h3
    class="relative text-xl mt-4"
    id="this-is-a-test"
  >
    this is a test
    <a
      class="anchor"
      href="#this-is-a-test"
    >
      ยง
    </a>
  </h3>
</div>
`;

exports[`renders H4 unchanged 1`] = `
<div>
  <h4
    class="relative text-xl mt-4"
    id="this-is-a-test"
  >
    this is a test
    <a
      class="anchor"
      href="#this-is-a-test"
    >
      ยง
    </a>
  </h4>
</div>
`;
